### ISSUE TITLE
Ruby CallCredentials does not invoke provided proc

### ISSUE THREAD

**Issue by anotherminh** [grpc/grpc#10665]
I'm trying to implement (non-Google) OAuth with `CallCredentials`, like so:
```
GRPC::Core::CallCredentials.new(
  proc do
    puts "generating new credentials!"
    { auth_header: "blahblahblah" }
  end
)
```

This doesn't work (the server doesn't see the auth_header that it expects). It seems that the proc is never called, so how are the credentials being passed along?  Maybe this isn't the right way to use CallCredentials.  I just couldn't find much documentation for supporting custom authentication for the ruby gRPC library.  Any help is much appreciated!

**Comment by apolcyn** [comment]
Are the call credentials being used on a secure channel? e.g., using `ChannelCredentials`?

There's some documentation (with ruby and other languages) in http://www.grpc.io/docs/guides/auth.html, with an example use of `CallCredentials` in the "Authenticate with Google" section

**Comment by anotherminh** [comment]
Does it ONLY work when composed with ChannelCredentials?  I've played with the Java grpc library, and for that library, CallCredentials work without ChannelCredentials.  Here's how it's being used in ruby:

```
creds = GRPC::Core::CallCredentials.new(proc { auth_header: "blahblahblah" })
stub = NameSpace::Stub.new("someaddress", :this_channel_is_insecure)
stub.get_foo(request_object, credentials: creds)
```

**Comment by apolcyn** [comment]
yes AFAIK this is the case with the ruby client

**Comment by anotherminh** [comment]
I see. Thanks for your help!

**Comment by colinmorelli** [comment]
Is there any chance this could be revisited? Running into the same issue now with Node (I would assume anything based on the C API?)

This is likely to become more of an issue if service meshes such as Istio rise in popularity, as they offload mTLS outside of the client applications anyway (thus making the RPC appear insecure to the app, while still going secure over the wire).

It's not the end of the day in the interim, it just means that I have to craft `Metadata` instances to pass around manually, but it would be _much_ cleaner if gRPC would allow this. If there's concern over the security of passing credentials on insecure channels, what about an option that is disabled by default?

**Comment by apolcyn** [comment]
> If there's concern over the security of passing credentials on insecure channels, what about an option that is disabled by default?

reopening this issue to look into this last suggestion

**Comment by colinmorelli** [comment]
Thanks! I posted over on the grpc-io mailing list [here](https://groups.google.com/forum/#!topic/grpc-io/ifL63H0kN48) some reasons why this makes sense to me. It feels worth enumerating them again here if this is where the discussion would happen:

1) At least gRPC-java does support this. At best, the inconsistency is strange, at worst it could lead to painful realizations down the road if starting on Java and assuming that similar patterns will "just work" on other languages. This is what happened in my case, where our gRPC-Java implementations worked fine, but attempting to do the same thing in Node did not work and took a while before I realized this was the reason.
2) While I understand gRPC's belief that it's insecure to exchange tokens over plaintext channels, the reality is that the application-level implementation really has no idea what channel the data will actually be exchanged over. For example, in Istio deployments, the application may think it's communicating insecurely (and thus not allow CallCredentials to be sent), when in fact the traffic is going to hit an external container that will perform mTLS auth with the destination service. From the client and server perspective, this is an insecure channel, but in reality - it's not (unless you're concerned about the ability to tcpdump the loopback interface - at which point you're probably screwed anyway).
3) There are plenty of cases where the CallCredentials themselves are not nearly as sensitive (i.e. they're enforced single-use tokens like JWTs), and thus may be less risky to exchange over plaintext anyway. This could be the case in scenarios where the services themselves are not dealing with private information, but perhaps they perform an action that should still be authenticated. Understandably, everything should be TLS anyway, but see point #2 for cases in which the service could be using TLS in ways that gRPC may not know about.
4) Finally, from a developer experience perspective, it's still possible to send this information anyway - but it results in more fragile implementations of gRPC clients. For example, in Node, I've worked around this limitation by simply pre-generating Metadata instances that can be passed to calls (instead of using CallCredentials), but this requires me to take care to ensure that, at all call-sites, I have valid metadata (i.e. it hasn't expired since it was generated - we're using JWTs). CallCredentials provide a single way for me to do this, but it's currently not possible because of the restriction to use secure channels.
